core/src/main/scala/org/bitcoins/core/wallet/utxo/UTXOSpendingInfo.scala

```diff
@@ -17,26 +40,245 @@ sealed abstract class UTXOSpendingInfo {
+                  case Some(witness: ScriptWitnessV0) => Some(witness)
+                  case None                           => None
+                  case Some(_: ScriptWitness) =>
+                    throw new UnsupportedOperationException(
+                      "Only v0 Segwit is currently supported")
+                }
+                P2SHNestedSegwitV0UTXOSpendingInfo(
+                  outPoint,
+                  output.value,
+                  p2sh,
+                  signers,
+                  hashType,
+                  wspk,
+                  witnessOpt.getOrElse(throw new IllegalArgumentException(
+                    "Script Witness must be defined for (nested) Segwit input"))
+                )
+              case _: ScriptPubKey =>
+                P2SHSpendingInfo(outPoint,
+                                 output.value,
+                                 p2sh,
+                                 signers,
+                                 hashType,
+                                 redeemScript)
+            }
+        }
+      case wspk: WitnessScriptPubKeyV0 =>
+        val witnessOpt = scriptWitnessOpt match {
+          case Some(witness: ScriptWitnessV0) => Some(witness)
+          case None                           => None
+          case Some(_: ScriptWitness) =>
+            throw new UnsupportedOperationException(
+              "Only v0 Segwit is currently supported")
+        }
+
+        SegwitV0NativeUTXOSpendingInfo(
+          outPoint,
+          output.value,
+          wspk,
+          signers,
+          hashType,
+          witnessOpt.getOrElse(
+            throw new IllegalArgumentException(
+              "Script Witness must be defined for Segwit input"))
+        )
+      case wspk: UnassignedWitnessScriptPubKey =>
+        UnassignedSegwitNativeUTXOSpendingInfo(
+          outPoint,
+          output.value,
+          wspk,
+          signers,
+          hashType,
+          scriptWitnessOpt.getOrElse(EmptyScriptWitness))
+      case _: P2PKScriptPubKey | _: P2PKHScriptPubKey |
+          _: MultiSignatureScriptPubKey | _: NonStandardScriptPubKey |
+          _: CLTVScriptPubKey | _: CSVScriptPubKey | _: WitnessCommitment |
+          EmptyScriptPubKey =>
+        RawScriptUTXOSpendingInfo(outPoint,
+                                  output.value,
+                                  output.scriptPubKey,
+                                  signers,
+                                  hashType)
+    }
+  }
+
+  // TODO: Get rid of this and force all matches to match on the ADT
+  def unapply(info: BitcoinUTXOSpendingInfo): Option[
+    (
+        TransactionOutPoint,
+        TransactionOutput,
+        Seq[Sign],
+        Option[ScriptPubKey],
+        Option[ScriptWitness],
+        HashType)] = {
+    Some(info.outPoint,
+         info.output,
+         info.signers,
+         info.redeemScriptOpt,
+         info.scriptWitnessOpt,
+         info.hashType)
+  }
+
+  def validateScriptWitness(
```

Comment:
nit: I would prefer this to be `isValidScriptWitness(): Boolean` and then do
```scala
require(isValidScriptWitness(params...), "message....") 
```

just to keep with convention. This looked weird to me when reading our code and i don't believe we use it anywhere else

```scala
BitcoinUTXOSpendingInfo.validateScriptWitness(redeemScript, scriptWitness)
```